app/i18n.py

```diff
@@ -0,0 +1,64 @@
+"""i18n handling
+"""
+import contextlib
+import contextvars
+import gettext
+import logging
+
+from .settings import I18N_DIR
+
+log = logging.getLogger(__name__)
+
+
+DEFAULT_LANGUAGE = "en"
+
+
+# a cache of loaded translations
+_translations = {}
+
+
+def get_translation(lang):
+    """Load a translations object, caching them"""
+    if lang not in _translations:
+        lang_translations = gettext.translation(
+            "knowledge-panel",
+            localedir=str(I18N_DIR),
+            languages=[lang, DEFAULT_LANGUAGE],
+        )
+        _translations[lang] = lang_translations
+    return _translations[lang]
+
+
+# current language context vars manage the current language
+_current_language = contextvars.ContextVar("current_language", default=None)
+
+
+@contextlib.contextmanager
+def active_translation(lang=None):
+    """A context manager to set the language"""
+    if lang is None:
+        lang = DEFAULT_LANGUAGE
+    # set current language
+    token = _current_language.set(lang)
+    yield
+    # restore context
+    _current_language.reset(token)
+
+
+def get_current_translation():
+    """function to get translation object for current language"""
+    lang = _current_language.get()
+    if lang is None:
+        # warn
+        log.warning(
+            "Looking up a translation while language was not set. "
+            "Use active_translation context manager",
+            stack_info=True,
+        )
```

Comment:
@alexgarel Just a question what does `stack_info=True` do?

Reply:
it print the stack trace along with the log message (like for exceptions).
Here I added it, because if I just log the developer won't see where the translation function was called, whereas here he see the lines.

Eg:

```
knowledge-panel | Looking up a translation while language was not set. Use active_translation context manager
knowledge-panel | Stack (most recent call last):
knowledge-panel |   File "/usr/local/lib/python3.9/threading.py", line 937, in _bootstrap
knowledge-panel |     self._bootstrap_inner()
knowledge-panel |   File "/usr/local/lib/python3.9/threading.py", line 980, in _bootstrap_inner
knowledge-panel |     self.run()
knowledge-panel |   File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run
knowledge-panel |     result = context.run(func, *args)
knowledge-panel |   File "/code/./app/main.py", line 45, in knowledge_panel
knowledge-panel |     panels.append(last_edits_kp(facet=facet_tag, value=value_tag, country=country))
knowledge-panel |   File "/code/./app/knowledge_panels.py", line 121, in last_edits_kp
knowledge-panel |     expected_html, t_description, t_title = last_edit(url=url, query=query)
knowledge-panel |   File "/code/./app/off.py", line 80, in last_edit
knowledge-panel |     description = _("last-edits issues related to")
knowledge-panel |   File "/code/./app/i18n.py", line 66, in translate
knowledge-panel |     return get_current_translation().gettext(message)
knowledge-panel |   File "/code/./app/i18n.py", line 55, in get_current_translation
knowledge-panel |     log.warning(
```

So you see the problem is in off.py line 80

Reply:
Remove the activate_translation line in main.py :-)